src/linear_algebra/general_linear_group.lean

```diff
@@ -0,0 +1,169 @@
+def det : GL n R →* units R :=
+{ to_fun := λ A,
+  { val := (↑A : matrix n n R).det,
+    inv := (↑(A⁻¹) : matrix n n R).det,
+    val_inv := by rw [←det_mul, ←mul_eq_mul, A.mul_inv, det_one],
+    inv_val := by rw [←det_mul, ←mul_eq_mul, A.inv_mul, det_one]},
+  map_one' := units.ext det_one,
+  map_mul' := λ A B, units.ext $ det_mul _ _ }
+
+/--The `GL n R` and `general_linear_group R n` groups are multiplicatively equivalent-/
+def to_lin : (GL n R) ≃* (linear_map.general_linear_group R (n → R)) :=
+units.map_equiv to_lin_alg_equiv'.to_mul_equiv
+
+/--Given a matrix with invertible determinant we get an element of `GL n R`-/
+def mk' (A : matrix n n R) (h : invertible (matrix.det A)) : GL n R :=
+unit_of_det_invertible A
+
+/--Given a matrix with unit determinant we get an element of `GL n R`-/
+noncomputable def mk'' (A : matrix n n R) (h : is_unit (matrix.det A)) : GL n R :=
+nonsing_inv_unit A h
+
+instance coe_fun : has_coe_to_fun (GL n R) :=
+{ F   := λ _, n → n → R,
+  coe := λ A, A.val }
+
+lemma ext_iff (A B : GL n R) : A = B ↔ (∀ i j, (A : matrix n n R) i j = (B : matrix n n R) i j) :=
+units.ext_iff.trans matrix.ext_iff.symm
+
+/-- Not marked `@[ext]` as the `ext` tactic already solves this. -/
+lemma ext ⦃A B : GL n R⦄ (h : ∀ i j, (A : matrix n n R) i j = (B : matrix n n R) i j) :
+  A = B :=
+units.ext $ matrix.ext h
+
+section coe_lemmas
+
+variables (A B : GL n R)
+
+@[simp] lemma coe_fn_eq_coe : ⇑A = (↑A : matrix n n R) := rfl
+
+@[simp] lemma mul_val : ↑(A * B) = A ⬝ B := rfl
+
+@[simp] lemma one_val : ↑(1 : GL n R) = (1 : matrix n n R) := rfl
+
+@[simp] lemma coe_inv : ↑(A⁻¹) = (↑A : matrix n n R)⁻¹ :=
+begin
+  letI := A.invertible,
+  exact inv_eq_nonsing_inv_of_invertible (↑A : matrix n n R),
+end
+
+@[simp] lemma to_lin'_mul : to_lin' (A * B) = (to_lin' A).comp (to_lin' B) :=
+matrix.to_lin'_mul A B
+
+@[simp] lemma to_lin'_one : to_lin' (1 : GL n R).1 = linear_map.id :=
+matrix.to_lin'_one
+
+end coe_lemmas
+
+instance SL_to_GL : has_coe (special_linear_group n R) (GL n R) :=
+⟨λ A, ⟨↑A, ↑(A⁻¹), congr_arg coe (mul_right_inv A), congr_arg coe (mul_left_inv A)⟩⟩
+
+end general_linear_group
+
+namespace GL_plus
+
+variables  (n : Type u) [decidable_eq n] [fintype n] (R : Type v) [linear_ordered_comm_ring R ]
+
+lemma one_in_GL_pos : 0 < det (1 : GL n R) :=
+begin
+simp only [det_one, general_linear_group.coe_fn_eq_coe, general_linear_group.one_val, zero_lt_one],
+end
+
+lemma mul_in_GL_pos (A B : GL n R) (h1 : 0 < det A) (h2 : 0 < det B) : 0 < det (A*B) :=
+begin
+  simp only [gt_iff_lt, det_mul, mul_eq_mul],
+  apply mul_pos h1 h2,
+end
+
+lemma inv_det_pos  (A : GL n R) (h : 0 < det A) :  0 < det (A⁻¹).1 :=
+begin
+  have h1 := is_unit_nonsing_inv_det A,
+  have h2 := nonsing_inv_det A,
+  have h3 : 0 < det ⇑A*  det (⇑A)⁻¹ , by {rw mul_comm, rw h2, linarith, simp,sorry,},
+  unfold_coes at *,
+  convert (zero_lt_mul_left h).mp h3,
+  have := general_linear_group.coe_inv A,
+  unfold_coes  at this,
+  simp_rw this,
+end
```

Comment:
This proof works:
```
  have : (0 : R) < (A⁻¹ * A).det := by simp,
  rw [monoid_hom.map_mul, units.coe_mul] at this,
  exact (zero_lt_mul_left h).mp this,
end
```

Reply:
this is giving a type mismatch

Reply:
Oh you're right, I cheated and changed the statement to something easier:
```lean
-- this is `general_linear_group.det` which is always a unit!
lemma inv_det_pos (A : GL n R) (h : (0 : R) < (A⁻¹).det) : (0 : R) < A.det :=
```

Reply:
Yeah, but with this one now the definition of GL_pos breaks at `inv_mem`. Should the it be ``lemma inv_det_pos (A : GL n R) (h : (0 : R) < A.det) : (0 : R) < (A⁻¹).det :=``

Reply:
I think so, so i made it that and fixed the proof.